### PR TITLE
MGMT-9085 Add vSphere Connection plugin

### PR DESCRIPTION
**Pull request by mareklibra**
[vSphere Plugin walk-through](https://www.youtube.com/watch?v=nwnUh3O4pmc)

### CODE REVIEW

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+      <li>
+        <Trans i18nKey="vsphere-plugin~vsphere-connection-form-folderhelp-one">
+          Provide <b>datacenter</b> folder which contains VMs of the cluster, example: /
+          <span className="vsphere-connection-form-helper__datacenter">{{ datacenter }}</span>/
+          <b>vm</b>/<b>[MY_VMS_TOP_FOLDER]</b>.
+        </Trans>
+      </li>
+      <li>
+        <Trans i18nKey="vsphere-plugin~vsphere-connection-form-folderhelp-two">
+          The file backing PersistenVolume will be stored with this prefix under the{' '}
```
Comment (jinyunma) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:95]: PersistenVolume -> PersistentVolume

**Comment by yapei** [comment]
it looks like all fields are marked as required (*) ,however user still can click`Save configuration` and resources got updated when any requirement is missing

**Review by jinyunma — COMMENTED** [review]
some comments about helpinfo

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  "Can be both domain name or IP, see additional info how to get it.": "Can be both domain name or IP, see additional info how to get it.",
+  "Username": "Username",
+  "Password": "Password",
+  "The password will be stored in a Secret in the kube-system namespace of this cluster.": "The password will be stored in a Secret in the kube-system namespace of this cluster.",
+  "Datacenter": "Datacenter",
+  "The name of an existing datacenter in the vSphere which the virtual machines backing this cluster are in.": "The name of an existing datacenter in the vSphere which the virtual machines backing this cluster are in.",
+  "Default data store": "Default data store",
+  "The name of an existing datastore in the datacenter where the persistent volumes will be stored.": "The name of an existing datastore in the datacenter where the persistent volumes will be stored.",
+  "Make sure there is /kubevols folder created in the root of the datastore.": "Make sure there is /kubevols folder created in the root of the datastore.",
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:30]: I would rephrase to cover both cases.
This plugin is focused especially on users with limited experience. Like me coming to it at the very beginning. And I had very hard time to get it working.

File: frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+const ConnectionFormContext = React.createContext<ConnectionFormContextData | null>(null);
+
+export const ConnectionFormContextProvider: React.FC<{
+  children: React.ReactNode;
+}> = ({ children }) => {
+  const [vcenter, setVcenter] = React.useState<string>('');
+  const [username, setUsername] = React.useState<string>('');
+  const [password, setPassword] = React.useState<string>('');
+  const [datacenter, setDatacenter] = React.useState<string>('');
+  const [defaultdatastore, setDefaultdatastore] = React.useState<string>('');
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx:13]:
```suggestion
  const [defaultDatastore, setDefaultDatastore] = React.useState<string>('');
```

File: frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        setFolder(v);
+      },
+    }),
+    [datacenter, defaultdatastore, folder, isDirty, password, username, vcenter],
+  );
+
+  return <ConnectionFormContext.Provider value={value}>{children}</ConnectionFormContext.Provider>;
+};
+
+export const useConnectionFormContext = () => {
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx:64]: why using context here ? It can be just a simple useConectionForm hook like 
```ts
export const useConnectionForm = (): ConnectionFormContextData => {
  const [vcenter, setVcenter] = React.useState<string>('');
  const [username, setUsername] = React.useState<string>('');
  const [password, setPassword] = React.useState<string>('');
  const [datacenter, setDatacenter] = React.useState<string>('');
  const [defaultdatastore, setDefaultdatastore] = React.useState<string>('');
  const [folder, setFolder] = React.useState<string>('');
  const [isDirty, setDirty] = React.useState(false);

  const callbacks = React.useMemo(() => ({
    setVcenter: (v) => {
      setDirty(true);
      setVcenter(v);
    },
    setUsername: (v) => {
      setDirty(true);
      setUsername(v);
    },
    setPassword: (v) => {
      setDirty(true);
      setPassword(v);
    },
    setDatacenter: (v) => {
      setDirty(true);
      setDatacenter(v);
    },
    setDefaultdatastore: (v) => {
      setDirty(true);
      setDefaultdatastore(v);
    },
    setFolder: (v) => {
      setDirty(true);
      setFolder(v);
    },
  }), []);

  return {
    vcenter,
    username,
    password,
    datacenter,
    defaultdatastore,
    folder,
    isDirty,
    setDirty,
    ...callbacks,
  }
};
```

The context makes it more complicated that it needs to be IMO

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.css [hunk 5]
```diff
@@ -0,0 +1,7 @@
+.vsphere-connection-form-helper__datacenter {
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.css:1]: lets align the css rules naming.
In previous css you're using a different prefix
```
.plugin-vsphere-status-popup
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+import './VSphereConnectionForm.css';
+
+export const VSphereConnectionForm: React.FC<VSphereConnectionProps & { formId?: string }> = ({
+  cloudProviderConfig,
+  formId,
+}) => {
+  const { t } = useTranslation();
+  const [isLoaded, setIsLoaded] = React.useState(false);
+  const [SecretModel] = useK8sModel({ group: 'app', version: 'v1', kind: 'Secret' });
+  const vcenterRef = React.useRef(null);
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:19]:
```suggestion
  const vcenterRef = React.useRef<HTMLInputElement>(null);
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    setPassword,
+    setDatacenter,
+    setDefaultdatastore,
+    setFolder,
+  } = useConnectionFormContext();
+
+  React.useEffect(() => {
+    if (vcenterRef && vcenterRef.current) {
+      (vcenterRef.current as HTMLInputElement).focus();
+    }
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:41]:
```suggestion
    vcenterRef?.current?.focus();
```

Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:41]:
```suggestion
    vcenterRef?.current?.focus();
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+import './VSphereConnectionForm.css';
+
+export const VSphereConnectionForm: React.FC<VSphereConnectionProps & { formId?: string }> = ({
+  cloudProviderConfig,
+  formId,
+}) => {
+  const { t } = useTranslation();
+  const [isLoaded, setIsLoaded] = React.useState(false);
+  const [SecretModel] = useK8sModel({ group: 'app', version: 'v1', kind: 'Secret' });
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:18]:
```suggestion
  const [secretModel] = useK8sModel({ group: 'app', version: 'v1', kind: 'Secret' });
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    const doItAsync = async () => {
+      if (isLoaded) {
+        return;
+      }
+      if (!cloudProviderConfig) {
+        return;
+      }
+
+      setIsLoaded(true);
+      await initialLoad(
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:55]: `initialLoad` is returning boolean but its not used anywhere

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+        label={t('vsphere-plugin~vCenter')}
+        labelIcon={
+          <PopoverHelpButton
+            content={
+              <>
+                {t(
+                  "vsphere-plugin~Enter the network address the vCenter is running on. It can either be a domain name or IP. If you're unsure, you can try to determine the value from the vSphere Web Client address. Example: ",
+                )}
+                <ul>
+                  <li> https://[your_vCenter_address]/ui</li>
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:137]: or maybe the space is there on purpose ?
```suggestion
                  <li>https://[your_vCenter_address]/ui</li>
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+import { PopoverHelpButton } from '../PopoverHelpButton';
+import { useConnectionFormContext } from './ConnectionFormContext';
+import { initialLoad } from './initialLoad';
+import { VSphereConnectionProps } from './types';
+
+import './VSphereConnectionForm.css';
+
+export const VSphereConnectionForm: React.FC<VSphereConnectionProps & { formId?: string }> = ({
+  cloudProviderConfig,
+  formId,
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:14]: why is `formId` passed as prop ? It does not look like something which would change

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+import { useK8sModel } from '@console/dynamic-plugin-sdk/src/api/core-api';
+import { useConnectionFormContext } from './ConnectionFormContext';
+import { InProgress } from './InProgress';
+import { persist } from './persist';
+import { VSphereConnectionProps } from './types';
+import { VSphereConnectionForm } from './VSphereConnectionForm';
+import { VSphereOperatorStatuses } from './VSphereOperatorStatuses';
+import './VSphereConnectionModal.css';
+
+export const VSphereConnectionModal: React.FC<VSphereConnectionProps> = (params) => {
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:23]:
```suggestion
export const VSphereConnectionModal: React.FC<VSphereConnectionProps> = (props) => {
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+    setDirty,
+  } = useConnectionFormContext();
+
+  const formId = 'vsphere-connection-modal-form';
+
+  const onClose = () => {
+    setModalOpen(false);
+
+    // hide popup
+    params.hide && params.hide();
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:50]:
```suggestion
    params.hide?.();
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+        },
+        params.cloudProviderConfig,
+      );
+
+      // Done
+      setIsSaving(false);
+
+      if (errorMsg) {
+        setError(errorMsg);
+        setIsSaving(false);
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:77]: this is unnecessary as its already called on line 73
```suggestion
```

File: frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    };
+
+    doItAsync();
+  }, [pollingTimmer, t]);
+
+  const onToggle = (value: boolean) => {
+    setIsExpanded(value);
+  };
+
+  // TODO: calculate from all vSphere-related operators
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx:136]: it this still not done ?

File: frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+      setKubeControllerManager(await getOperatorHealth(t, 'kube-controller-manager'));
+      setKubeApiServer(await getOperatorHealth(t, 'kube-apiserver'));
+      setStorage(await getOperatorHealth(t, 'storage'));
+
+      await delay(DELAY_BEFORE_POLLING_RETRY_MEDIUM);
+      setPollingTimmer(pollingTimmer + 1);
+    };
+
+    doItAsync();
+  }, [pollingTimmer, t]);
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx:130]: why are you polling instead of using `useK8sWatchResource` ?

File: frontend/packages/vsphere-plugin/src/components/utils.ts [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+        return `[VirtualCenter "${vcenter}"]`;
+      }
+      return line;
+    })
+    .join('\n');
+
+  return result;
+};
+
+export const waitForClusterOperator = async (
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/utils.ts:86]: this does not seem to be used anywhere

File: frontend/packages/vsphere-plugin/src/flag.ts [hunk 18]
```diff
@@ -0,0 +1,40 @@
+import * as React from 'react';
+import { SetFeatureFlag } from '@console/dynamic-plugin-sdk';
+import { k8sGet, useK8sModel } from '@console/dynamic-plugin-sdk/src/api/core-api';
+import { VSPHERE_FEATURE_FLAG, VSPHERE_PLATFORM } from './constants';
+import { Infrastructure } from './resources';
+
+export const useVSphereFlagHandler = (setFeatureFlag: SetFeatureFlag) => {
+  // eslint-disable-next-line no-console
+  console.log('useVSphereFlagHandler called', setFeatureFlag);
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/flag.ts:9]: lets remove the unnecessary logging

File: frontend/packages/vsphere-plugin/src/plugin.ts [hunk 19]
```diff
@@ -0,0 +1,2 @@
+// See console-extensions.json instead
+export default [];
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/plugin.ts:2]: is this file needed ?

File: frontend/packages/vsphere-plugin/src/resources/statusCondition.ts [hunk 20]
```diff
@@ -0,0 +1,20 @@
+import { K8sResourceCondition } from '@console/internal/module/k8s';
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/resources/statusCondition.ts:1]: I think the only console imports should be from sdk package, not others

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+import { VSphereConnectionForm } from './VSphereConnectionForm';
+import { VSphereOperatorStatuses } from './VSphereOperatorStatuses';
+import './VSphereConnectionModal.css';
+
+export const VSphereConnectionModal: React.FC<VSphereConnectionProps> = (params) => {
+  const { t } = useTranslation();
+  const [isModalOpen, setModalOpen] = React.useState(true);
+
+  const [SecretModel] = useK8sModel({ group: 'app', version: 'v1', kind: 'Secret' });
+  const [ConfigMapModel] = useK8sModel({ group: 'app', version: 'v1', kind: 'ConfigMap' });
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:28]:
```suggestion
  const [secretModel] = useK8sModel({ group: 'app', version: 'v1', kind: 'Secret' });
  const [configMapModel] = useK8sModel({ group: 'app', version: 'v1', kind: 'ConfigMap' });
```

File: frontend/packages/vsphere-plugin/src/components/VSphereConnection.tsx [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+import { VSphereConnectionModal } from './VSphereConnectionModal';
+
+const setPopupVisibility = (v: 'visible' | 'hidden') => {
+  const popup = document.getElementsByClassName('plugin-vsphere-status-popup');
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const style = (popup?.[0] as any)?.style;
+  style && (style.visibility = v);
+};
+
+export const VSphereConnection: React.FC<VSphereConnectionProps> = (params) => {
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnection.tsx:13]:
```suggestion
export const VSphereConnection: React.FC<VSphereConnectionProps> = (props) => {
```

File: frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.css [hunk 23]
```diff
@@ -0,0 +1,4 @@
+.plugin-vsphere-status-popup {
+    max-width: 30rem !important;
+    width: 30rem;
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.css:3]: I think it does not make sense to define both ? `width` should be enough, right ?

**Review by rawagner — CHANGES_REQUESTED** [review]

File: frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.tsx [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+  const health = getVSphereHealth(t, responses, additionalResource);
+  const { state } = health;
+
+  let message: string | undefined;
+  switch (state) {
+    case HealthState.WARNING:
+      message = health.message;
+      break;
+    case HealthState.PROGRESS:
+      message = t('vsphere-plugin~Click for more details');
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.tsx:62]: Because we do not know...
The backend is not reporting progress and it's difficult to "guess" it from conditions of related resources (like nodes or operators).
After a lot of discussions we came to the conclusion to leave that on the user until improved at the backend. And this is one of the less controvesion phrase of considered ones.

File: frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.tsx [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+const VSphereStatus: React.FC<PrometheusHealthPopupProps> = ({ hide, responses, k8sResult }) => {
+  const { t } = useTranslation();
+  const health = getVSphereHealth(t, responses, k8sResult);
+
+  if (
+    [HealthState.OK, HealthState.WARNING, HealthState.PROGRESS].includes(health.state) &&
+    k8sResult?.data
+  ) {
+    const cloudProviderConfig = k8sResult.data as ConfigMap | undefined;
```
[the remark this answers is not in the document]
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.tsx:26]: done

File: frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx (hunk 4 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx:64]: Leftover, the flow should be more complex and the context would help.
When simplified to recent state, I kept the context when it has been already implemented.
I can remove it as well now if needed. On the other hand, we might add additional screens later and so will be probably implementing context again.

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx (hunk 10 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:137]: done

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx (hunk 14 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:77]: right

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx (hunk 13 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:50]: ok, done

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx (hunk 12 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:23]: done

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx (hunk 21 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:28]: done

File: frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx (hunk 3 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/ConnectionFormContext.tsx:13]: done

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx (hunk 6 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.tsx:19]: done

File: frontend/packages/vsphere-plugin/src/components/utils.ts (hunk 17 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/utils.ts:86]: Right, not anymore. Removing.

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.css (hunk 5 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionForm.css:1]: done

File: frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx (hunk 15 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx:136]: yes, it is

File: frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx (hunk 16 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx:130]: done

File: frontend/packages/vsphere-plugin/src/plugin.ts (hunk 19 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/plugin.ts:2]: Every other plugin has an entrypoint.

File: frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.css (hunk 23 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.css:3]: This line is overriding it: https://github.com/openshift/console/blob/master/frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx#L58

Even with `!important`, the result is not as expected when just a single attribute is set here.

Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/ClusterOverview/VSphereStatus.css:3]: We can make it more generic there but this is fast workaround. I can add a comment to fix later.

File: frontend/packages/vsphere-plugin/src/resources/model.ts [hunk 26]
```diff
@@ -0,0 +1,14 @@
+import { K8sModel } from '@console/dynamic-plugin-sdk';
+
+// /apis/operator.openshift.io/v1/kubecontrollermanagers/cluster'
+export const KubeControllerManagerModel: K8sModel = {
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/resources/model.ts:4]: you should use `useK8sModel` instead of this static model definition

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+    password?.trim() &&
+    datacenter?.trim() &&
+    defaultDatastore?.trim();
+  const isSaveDisabled = isSaving || !isDirty || !allRequiredFieldsAreSet;
+
+  const footer = (
+    <>
+      {isSaving ? <InProgress key="progress" text={t('vsphere-plugin~Saving...')} /> : null}
+      <Button key="cancel" variant="link" onClick={onClose}>
+        Cancel
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:137]: missing i18n

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+      </Button>
+      <Button key="save" variant="primary" isDisabled={isSaveDisabled} onClick={onSave}>
+        Save configuration
```
Comment (rawagner) [inline frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:140]: missing i18n

File: frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx (hunk 16 above)
Comment (rawagner) [reply frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx]: I dont see any chage here, did you forget to commit ?

**Review by rawagner — CHANGES_REQUESTED** [review]

File: frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx (hunk 16 above)
Comment (rawagner) [reply frontend/packages/vsphere-plugin/src/components/VSphereOperatorStatuses.tsx]: old diff, I see it now ! :)

File: frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx (hunk 27 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/components/VSphereConnectionModal.tsx:137]: done

File: frontend/packages/vsphere-plugin/src/resources/model.ts (hunk 26 above)
Comment (mareklibra) [reply frontend/packages/vsphere-plugin/src/resources/model.ts:4]: done

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+  "Username": "Username",
+  "Password": "Password",
+  "The password will be stored in a Secret in the kube-system namespace of this cluster.": "The password will be stored in a Secret in the kube-system namespace of this cluster.",
+  "Datacenter": "Datacenter",
+  "Default data store": "Default data store",
+  "Virtual Machine Folder": "Virtual Machine Folder",
+  "vSphere Problem Detector (can be outdated)": "vSphere Problem Detector (can be outdated)",
+  "Retry": "Retry",
+  "Delayed propagation of configuration": "Delayed propagation of configuration",
+  "After saving the configuration, it may take approximately one hour to see if the settings are correct and the operators' statuses are updated, nodes will get rebooted.": "After saving the configuration, it may take approximately one hour to see if the settings are correct and the operators' statuses are updated, nodes will get rebooted.",
```
Comment (opayne1) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:35]: A style suggestion to match documentation. See "May" in IBM style https://www.ibm.com/docs/en/ibm-style?topic=word-usage#m 

```suggestion
  "After saving the configuration, it might take approximately one hour to see if the settings are correct and the operators' statuses are updated, nodes will get rebooted.": "After saving the configuration, it might take approximately one hour to see if the settings are correct and the operators' statuses are updated, nodes will get rebooted.",
```

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+  "Failed to create {{secret}} secret": "Failed to create {{secret}} secret",
+  "Failed to load kubecontrollermanager": "Failed to load kubecontrollermanager",
+  "Failed to patch kubecontrollermanager": "Failed to patch kubecontrollermanager",
+  "Failed to patch {{cm}}": "Failed to patch {{cm}}",
+  "Failed to create {{cm}} ConfigMap": "Failed to create {{cm}} ConfigMap",
+  "vsphere-connection-form-folderhelp-one": "Provide <1>datacenter</1> folder which contains VMs of the cluster, example: /<3>{{datacenter}}</3>/<5>vm</5>/<7>[MY_VMS_TOP_FOLDER]</7>.",
+  "vsphere-connection-form-datacenterhelp-one": "The name of an existing datacenter in the vSphere which the virtual machines backing this cluster are in.",
+  "vsphere-connection-form-datacenterhelp-two": "Please mind, that changing this value will break existing PersistentVolumes, if there are already any.",
+  "vsphere-connection-form-datastorehelp-one": "The name of an existing datastore in the datacenter where the persistent volumes will be stored.",
+  "vsphere-connection-form-datastorehelp-two": "Please mind, that changing this value will break existing PersistentVolumes, if there are already any.",
```
Comment (opayne1) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:22]: I would consider revising this line and line 20 to be more direct and use active rather than passive voice. Can we do  a "note" instead of "please mind"? So,
```suggestion
  "vsphere-connection-form-datastorehelp-two": "Note: Updating this value will break any existing PersistentVolumes.",
```
I would also suggest removing the "if there are already any" since you already said "existing".

Comment (opayne1) [reply frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:22]: or 
```suggestion
  "vsphere-connection-form-datastorehelp-two": "Note, changing this value will break existing PersistentVolumes.",
```

**Review by opayne1 — COMMENTED** [review]
Looks good! I left a few comments to consider.

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 31]
```diff
@@ -0,0 +1,54 @@
+{
+  "vSphere connection": "vSphere connection",
+  "vSphere configurations": "vSphere configurations",
+  "Loading vSphere connection status...": "Loading vSphere connection status...",
```
Comment (wolfeallison) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:4]: I think we could erase the ellipses (...) because ellipses are used when information is omitted, when you cannot fit all words onto a line, or when you remove less relevant information (like in a quote) https://www.patternfly.org/v4/ux-writing/punctuation

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+  "The vSphere Connection check is failing.": "The vSphere Connection check is failing.",
+  "Click for more details": "Click for more details",
```
Comment (wolfeallison) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:6]: Where is "Click for more details" going? I would recommend being more specific on what details you are providing and hyperlinking the sentence. We try to avoid saying "Click here" or anything related to make it more accessible for people using screen readers. I would suggest saying something like "Learn more about ___" with the whole sentence hyperlinked

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+  "Failing {{reason}}": "Failing {{reason}}",
+  "Failed to patch {{secret}}": "Failed to patch {{secret}}",
+  "Failed to create {{secret}} secret": "Failed to create {{secret}} secret",
+  "Failed to load kubecontrollermanager": "Failed to load kubecontrollermanager",
+  "Failed to patch kubecontrollermanager": "Failed to patch kubecontrollermanager",
+  "Failed to patch {{cm}}": "Failed to patch {{cm}}",
+  "Failed to create {{cm}} ConfigMap": "Failed to create {{cm}} ConfigMap",
+  "vsphere-connection-form-folderhelp-one": "Provide <1>datacenter</1> folder which contains VMs of the cluster, example: /<3>{{datacenter}}</3>/<5>vm</5>/<7>[MY_VMS_TOP_FOLDER]</7>.",
+  "vsphere-connection-form-datacenterhelp-one": "The name of an existing datacenter in the vSphere which the virtual machines backing this cluster are in.",
+  "vsphere-connection-form-datacenterhelp-two": "Please mind, that changing this value will break existing PersistentVolumes, if there are already any.",
```
Comment (wolfeallison) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:20]: The word "please" doesn't translate great in all languages (it can come across like optional). Instead of "Please mind" I would suggest "Note" or just starting right at "Changing this value..."

Comment (wolfeallison) [reply frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:20]: same with line 22

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json (hunk 29 above)
Comment (wolfeallison) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:35]: To match PatternFly's style guide, I'd suggest using numerics instead of writing out the number. So, "1" here instead of one

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 34]
```diff
@@ -0,0 +1,54 @@
...
+  "The password will be stored in a Secret in the kube-system namespace of this cluster.": "The password will be stored in a Secret in the kube-system namespace of this cluster.",
+  "Datacenter": "Datacenter",
+  "Default data store": "Default data store",
+  "Virtual Machine Folder": "Virtual Machine Folder",
+  "vSphere Problem Detector (can be outdated)": "vSphere Problem Detector (can be outdated)",
+  "Retry": "Retry",
+  "Delayed propagation of configuration": "Delayed propagation of configuration",
+  "After saving the configuration, it may take approximately one hour to see if the settings are correct and the operators' statuses are updated, nodes will get rebooted.": "After saving the configuration, it may take approximately one hour to see if the settings are correct and the operators' statuses are updated, nodes will get rebooted.",
+  "Note, that existing resources (like already bound PVCs) might get broken by these changes.": "Note, that existing resources (like already bound PVCs) might get broken by these changes.",
+  "Saving...": "Saving...",
```
Comment (wolfeallison) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:37]: Same comment here about ellipses

**Comment by mareklibra** [comment]
rebased

File: frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json [hunk 35]
```diff
@@ -0,0 +1,55 @@
...
+  "vsphere-connection-form-datastorehelp-one": "The name of an existing datastore in the datacenter where the persistent volumes will be stored.",
+  "vsphere-connection-form-datastorehelp-two": "Note: Updating this value will break any existing PersistentVolumes.",
+  "vCenter": "vCenter",
+  "Enter the network address the vCenter is running on. It can either be a domain name or IP. If you're unsure, you can try to determine the value from the vSphere Web Client address. Example: ": "Enter the network address the vCenter is running on. It can either be a domain name or IP. If you're unsure, you can try to determine the value from the vSphere Web Client address. Example: ",
+  "Can be both domain name or IP, see additional info how to get it.": "Can be both domain name or IP, see additional info how to get it.",
+  "Username": "Username",
+  "The vSphere Center username. Mind to double-check it is correct, otherwise cluster nodes will get unschedulable (Known issue: OCPBUGS-2353).": "The vSphere Center username. Mind to double-check it is correct, otherwise cluster nodes will get unschedulable (Known issue: OCPBUGS-2353).",
```
Comment (jinyunma) [inline frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json:27]: "The vSphere Center" -> "The vSphere vCenter"

Comment (mareklibra) [reply frontend/packages/vsphere-plugin/locales/en/vsphere-plugin.json]: fixed, thanks
